### PR TITLE
Disable FAv2.1+ for causal mask in cross attention

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR disables the use of FAv2.1+ for causal mask and cross attention due to the change of behavior in FA: https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -2134,6 +2135,13 @@ def forward(
         if not _flash_attn_2_available and self.num_gqa_groups != self.num_attention_heads:
             use_flash_attention = False
 
+        # Disable FAv2.1+ for causal mask in cross attention due to
+        # https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag
+        if (_flash_attn_2_1_plus
+            and attn_mask_type == "causal"
+            and max_seqlen_q != max_seqlen_kv):
+            use_flash_attention = False
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:2143]: I think a warning would be useful here pointing to the change in behavior in FA's README.

Comment (cyanguwa) [reply transformer_engine/pytorch/attention.py:2143]: Sure

**Comment by Marks101** [comment]
Hello @cyanguwa, if I interpret these changes correctly they effectively disable flash attention for self attention while doing incremental decoding (inference). In this scenario the condition `max_seqlen_q != max_seqlen_kv` applies as well.

**Comment by ptrendx** [comment]
@Marks101 Yes, I agree with you that this use case would benefit from the mew FA behavior (and it is an important usecase). The problem is that this introduces a (silent) behavior difference between this TE version and the previous ones, as well as a difference between FA and non-FA backends inside TE. We can't do that, and we are about to cut the release branch for v1.1, therefore we need to proceed with disabling FA in this case.

The way we think about solving it going forward is via changing our API a little bit to add the InferenceParams as the optional argument to the forward of DotProductAttention, rather than passing the full `k` and `v` with small `q`. This will enable us to figure out that we are in the decoding environment and infer the real shapes of q, k and v rather than overloading the cross attention semantics. This will also enable us then to support decoding with seq_length > 1 and be able to leave the `causal` mask on during inference, since there will be no ambiguity about the intent. Does that make sense to you @Marks101?

I'm going to merge this PR now and we will work towards this new approach in the coming weeks. @Marks101 Please don't hesitate to comment and critique the approach if you believe there are better ways to achieve this :-)!

**Review by ptrendx — APPROVED** [review]

**Comment by Marks101** [comment]
Hi @ptrendx, thank you so much for the elaboration on your decision. You are of course right to assure that the behavior of TE does not change with its versions.
The approach you suggested sounds reasonable. I guess I was just happy to see that with #506 we have the chance to bump flash-attn to fix our decoding issue (#512) and I wanted to put some awareness on this. But I am happy to hear that you have discussed it and I am looking forward to this solution. Thank you!

**Pull request merged**